sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,707 @@
+            b = self.table[b][A_dict[word.subword(j, j+1).inverse()]]
+            j -= 1
+        if j < i:
+            # we have an incorrect completed scan with coincidence f ~ b
+            # run the "coincidence" routine
+            self.coincidence(f, b)
+        elif j == i:
+            # deduction process
+            self.table[f][A_dict[word.subword(i, i+1)]] = b
+            self.table[b][A.A_dict[word.subword(i, i+1).inverse()]] = f
+        # otherwise scan is incomplete and yields no information
+
+    def merge(self, k, lamda, q):
+        p = self.p
+        phi = self.rep(k)
+        chi = self.rep(lamda)
+        if phi != chi:
+            mu = min(phi, chi)
+            v = max(phi, chi)
+            p[v] = mu
+            q.append(v)
+
+    def rep(self, k):
+        p = self.p
+        lamda = k
+        rho = p[lamda]
+        while rho != lamda:
+            lamda = rho
+            rho = p[lamda]
+        mu = k
+        rho = p[mu]
+        while rho != lamda:
+            p[mu] = lamda
+            mu = rho
+            rho = p[mu]
+        return lamda
+
+    # alpha, beta coincide
+    # here alpha, beta represent the pair of cosets where
+    # coincidence occurs
+    def coincidence(self, alpha, beta):
+        """
+        """
+        A_dict = self.A_dict
+        p = self.p
+        l = 0
+        # behaves as a queue
+        q = []
+        self.merge(alpha, beta, q)
+        while len(q) > 0:
+            gamma = q.pop(0)
+            # comment by Kalevi, this is already done by p[v] = mu
+            # del C[gamma]
+            # commenting this out
+            # omega = omega - set(gamma)
+            for x in A_dict:
+                delta = self.table[gamma][A_dict[x]]
+                if delta is not None:
+                    self.table[delta][A_dict[x**-1]] = None
+                    mu = self.rep(gamma)
+                    nu = self.rep(delta)
+                    if self.table[mu][A_dict[x]] is not None:
+                        self.merge(nu, self.table[mu][A_dict[x]], q)
+                    elif self.table[nu][A_dict[x**-1]] is not None:
+                        self.merge(mu, self.table[nu][A_dict[x**-1]], q)
+                    else:
+                        self.table[mu][A_dict[x]] = nu
+                        self.table[nu][A_dict[x**-1]] = mu
+
+    # method used in the HLT strategy
+    def scan_and_fill(self, alpha, word):
+        f = alpha
+        i = 0
+        r = len(word)
+        A_dict = self.A_dict
+        l_A = len(A_dict)
+        i_A = 0
+        while i_A < l_A:
+            # do the forward scanning
+            while i < r and self.table[f][A_dict[word.subword(i, i+1)]] is not None:
+                f = self.table[f][A_dict[word.subword(i, i+1)]]
+                i += 1
+            if i >= r:
+                if f != alpha:
+                    self.coincidence(f, alpha)
+                    return
```

Comment:
This `return` should probably be indented the same as `if f ...`. So we are finished in any case.